File: flutter_ui/lib/geom/slice9.dart
import 'rectangle.dart';

class Slice9Rects {
  final List<Rectangle> src, dst;
  Slice9Rects({required this.src, required this.dst});
}

class Slice9 {
  static Slice9Rects buildRects(double w, double h, double bitmapWidth,
      double bitmapHeight, Rectangle slice) {
    List<Rectangle> srcRects = buildSrcRects(bitmapWidth, bitmapHeight, slice);
    List<Rectangle> dstRects = buildDstRects(w, h, srcRects);
    return Slice9Rects(src: srcRects, dst: dstRects);
  }

  static List<Rectangle> buildSrcRects(
      double bitmapWidth, double bitmapHeight, Rectangle slice) {
    double x1 = slice.left, y1 = slice.top, x2 = slice.right, y2 = slice.bottom;
    return [
      Rectangle(0, 0, x1, y1),
      Rectangle(x1, 0, x2 - x1, y1),
      Rectangle(x2, 0, bitmapWidth - x2, y1),
      Rectangle(0, y1, x1, y2 - y1),
      Rectangle(x1, y1, x2 - x1, y2 - y1),
      Rectangle(x2, y1, bitmapWidth - x2, y2 - y1),
      Rectangle(0, y2, x1, bitmapHeight - y2),
      Rectangle(x1, y2, x2 - x1, bitmapHeight - y2),
      Rectangle(x2, y2, bitmapWidth - x2, bitmapHeight - y2)
    ];
  }

  static List<Rectangle> buildDstRects(
      double w, double h, List<Rectangle> srcRects) {
    return [
      Rectangle(0, 0, srcRects[0].width, srcRects[0].height),
      Rectangle(srcRects[0].width, 0, w - srcRects[0].width - srcRects[2].width,
          srcRects[1].height),
      Rectangle(
          w - srcRects[2].width, 0, srcRects[2].width, srcRects[2].height),
      Rectangle(0, srcRects[0].height, srcRects[3].width,
          h - srcRects[0].height - srcRects[6].height),
      Rectangle(
          srcRects[3].width,
          srcRects[0].height,
          w - srcRects[3].width - srcRects[5].width,
          h - srcRects[1].height - srcRects[7].height),
      Rectangle(w - srcRects[5].width, srcRects[2].height, srcRects[5].width,
          h - srcRects[2].height - srcRects[8].height),
      Rectangle(0, h - srcRects[6].height, srcRects[6].width, srcRects[6].height),
      Rectangle(srcRects[6].width, h - srcRects[7].height,
          w - srcRects[6].width - srcRects[8].width, srcRects[7].height),
      Rectangle(w - srcRects[8].width, h - srcRects[8].height, srcRects[8].width,
          srcRects[8].height)
    ];
  }
}

File: luaui/geom/Slice9.lua
local Rectangle = require("luaui.geom.Rectangle")
local Slice9 = {}

function Slice9.buildRects(w, h, bitmapWidth, bitmapHeight, slice)
    local srcRects = Slice9.buildSrcRects(bitmapWidth, bitmapHeight, slice)
    local dstRects = Slice9.buildDstRects(w, h, srcRects)
    return { src = srcRects, dst = dstRects }
end

function Slice9.buildSrcRects(bw, bh, slice)
    local x1, y1 = slice.left, slice.top
    local x2, y2 = slice:get_right(), slice:get_bottom()
    local r = {}
    table.insert(r, Rectangle.new(0, 0, x1, y1))
    table.insert(r, Rectangle.new(x1, 0, x2 - x1, y1))
    table.insert(r, Rectangle.new(x2, 0, bw - x2, y1))
    table.insert(r, Rectangle.new(0, y1, x1, y2 - y1))
    table.insert(r, Rectangle.new(x1, y1, x2 - x1, y2 - y1))
    table.insert(r, Rectangle.new(x2, y1, bw - x2, y2 - y1))
    table.insert(r, Rectangle.new(0, y2, x1, bh - y2))
    table.insert(r, Rectangle.new(x1, y2, x2 - x1, bh - y2))
    table.insert(r, Rectangle.new(x2, y2, bw - x2, bh - y2))
    return r
end

function Slice9.buildDstRects(w, h, src)
    local r = {}
    table.insert(r, Rectangle.new(0, 0, src[1].width, src[1].height))
    table.insert(r, Rectangle.new(src[1].width, 0, w - src[1].width - src[3].width, src[2].height))
    table.insert(r, Rectangle.new(w - src[3].width, 0, src[3].width, src[3].height))
    table.insert(r, Rectangle.new(0, src[1].height, src[4].width, h - src[1].height - src[7].height))
    table.insert(r, Rectangle.new(src[4].width, src[1].height, w - src[4].width - src[6].width, h - src[1].height - src[7].height))
    table.insert(r, Rectangle.new(w - src[6].width, src[1].height, src[6].width, h - src[1].height - src[7].height))
    table.insert(r, Rectangle.new(0, h - src[7].height, src[7].width, src[7].height))
    table.insert(r, Rectangle.new(src[7].width, h - src[8].height, w - src[7].width - src[9].width, src[8].height))
    table.insert(r, Rectangle.new(w - src[9].width, h - src[9].height, src[9].width, src[9].height))
    return r
end

return Slice9
